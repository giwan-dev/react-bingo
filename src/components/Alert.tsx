import React from 'react';
import ReactDOM from 'react-dom';
import styled from 'styled-components';

interface AlertProps {
  isVisible: boolean;
  message: string;
  onConfirm: React.MouseEventHandler<HTMLElement>;
}

const Overlay = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.4);
`;

const AlertContainer = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  position: absolute;
  top: 50%;
  left: 50%;
  margin: auto;
  width: 500px;
  height: 80px;
  border-radius: 5px;
  padding: 20px;
  background-color: white;
  transform: translate(-50%, -50%);
`;

const alertPortal = document.getElementById('alert-portal');
const stopPropagation = (e: React.MouseEvent) => e.stopPropagation();

export default class Alert extends React.Component<AlertProps> {
  el: HTMLElement|null = null;

  constructor(props: AlertProps) {
    super(props);

    this.el = document.createElement('div');
  }

  componentDidMount() {
    if (alertPortal !== null && this.el !== null) {
      alertPortal.appendChild(this.el);
    }

  }

  componentWillUnmount() {
    if (alertPortal !== null && this.el !== null) {
      alertPortal.removeChild(this.el);
    }
  }

  private renderAlert() {
    const { isVisible, message, onConfirm } = this.props;

    if (!isVisible) {
      return null;
    }

    return (
      <Overlay onClick={onConfirm}>
        <AlertContainer onClick={stopPropagation}>
          {message}
          <button
            type="button"
            onClick={onConfirm}
          >
            확인
          </button>
        </AlertContainer>
      </Overlay>
    );
  }

  public render() {
    if (this.el === null) {
      return null;
    }

    return ReactDOM.createPortal(
      this.renderAlert(),
      this.el,
    );
  }
}
