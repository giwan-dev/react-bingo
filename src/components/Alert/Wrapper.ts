import styled from 'styled-components';

const AlertWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  position: absolute;
  top: 50%;
  left: 50%;
  margin: auto;
  width: 500px;
  height: 80px;
  border-radius: 5px;
  padding: 20px;
  background-color: white;
  transform: translate(-50%, -50%);
`;

export default AlertWrapper;
