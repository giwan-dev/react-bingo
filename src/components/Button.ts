import COLOR from 'colors';
import styled from 'styled-components';

const Button = styled.button`
  margin: 0;
  height: 35px;
  border: 0;
  border-radius: 5px;
  padding: 8px 16px;
  color: white;
  background-color: ${COLOR.primary};
  cursor: pointer;
  transition: opacity ease-out 0.2s;

  &:hover {
    opacity: 0.7;
  }
`;

export default Button;
