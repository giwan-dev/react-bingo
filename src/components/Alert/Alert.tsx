import React from 'react';
import ReactDOM from 'react-dom';
import Button from '../Button';
import AlertOverlay from './Overlay';
import AlertWrapper from './Wrapper';

interface AlertProps {
  isVisible: boolean;
  message: string;
  onConfirm: React.MouseEventHandler<HTMLElement>;
}

const alertPortal = document.getElementById('alert-portal') || {
  appendChild: () => {},
  removeChild: () => {},
};
export const stopPropagation = (e: React.MouseEvent) => e.stopPropagation();

export default class Alert extends React.Component<AlertProps> {
  private el: HTMLElement;

  constructor(props: AlertProps) {
    super(props);

    this.el = document.createElement('div');
  }

  componentDidMount() {
    alertPortal.appendChild(this.el);
  }

  componentWillUnmount() {
    alertPortal.removeChild(this.el);
  }

  private renderAlert() {
    const { isVisible, message, onConfirm } = this.props;

    if (!isVisible) {
      return null;
    }

    return (
      <AlertOverlay onClick={onConfirm}>
        <AlertWrapper onClick={stopPropagation}>
          {message}
          <Button
            type="button"
            onClick={onConfirm}
          >
            확인
          </Button>
        </AlertWrapper>
      </AlertOverlay>
    );
  }

  public render() {
    if (this.el === null) {
      return null;
    }

    return ReactDOM.createPortal(
      this.renderAlert(),
      this.el,
    );
  }
}
