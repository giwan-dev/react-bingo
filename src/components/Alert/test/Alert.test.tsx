import React from 'react';
import Alert, { stopPropagation } from '../Alert';
import { shallow } from 'enzyme';
import Button from 'components/Button';
import AlertOverlay from '../Overlay';
import AlertWrapper from '../Wrapper';

describe('stopPropagation function', () => {
  it('should call stopPropagation property of input argument', () => {
    const property = jest.fn();
    const event = {
      stopPropagation: property,
    } as any;
    stopPropagation(event);
    expect(property).toHaveBeenCalled();
  });
});

describe('Alert component', () => {
  it('should render without crashing', () => {
    const alert = shallow((
      <Alert
        isVisible={true}
        message="hello, world"
        onConfirm={jest.fn()}
      />
    ));

    expect(alert).toMatchSnapshot();
  });

  it('should render message', () => {
    const message = 'hello, world';
    const alert = shallow((
      <Alert
        isVisible={true}
        message={message}
        onConfirm={jest.fn()}
      />
    ));

    expect(alert.contains(message)).toEqual(true);
  });

  it('should render Button component', () => {
    const alert = shallow((
      <Alert
        isVisible={true}
        message="hello, world"
        onConfirm={jest.fn()}
      />
    ));

    expect(alert.find(Button).exists()).toEqual(true);
  });

  it('should not render when isVisible is false', () => {
    const alert = shallow((
      <Alert
        isVisible={false}
        message="hello, world"
        onConfirm={jest.fn()}
      />
    ));

    expect(alert).toMatchSnapshot();
    expect(alert.children().exists()).toEqual(false);
  });

  it('should call onConfirm prop when click button', () => {
    const handleConfirm = jest.fn();
    const alert = shallow((
      <Alert
        isVisible={true}
        message="hello, world"
        onConfirm={handleConfirm}
      />
    ));

    alert.find(Button).simulate('click');
    expect(handleConfirm).toBeCalled();
  });

  it('should call onConfirm prop when click Overlay', () => {
    const handleConfirm = jest.fn();
    const alert = shallow((
      <Alert
        isVisible={true}
        message="hello, world"
        onConfirm={handleConfirm}
      />
    ));

    alert.find(AlertOverlay).simulate('click');
    expect(handleConfirm).toBeCalled();
  });

  it('should not call onConfirm prop when click Wrapper', () => {
    const handleConfirm = jest.fn();
    const alert = shallow((
      <Alert
        isVisible={true}
        message="hello, world"
        onConfirm={handleConfirm}
      />
    ));

    alert.find(AlertWrapper).simulate('click', new Event('click'));
    expect(handleConfirm).toHaveBeenCalledTimes(0);
  });
});
