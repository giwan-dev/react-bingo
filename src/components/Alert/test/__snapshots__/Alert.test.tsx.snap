// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert component should not render when isVisible is false 1`] = `
<Portal
  containerInfo={<div />}
/>
`;

exports[`Alert component should render without crashing 1`] = `
<Portal
  containerInfo={<div />}
>
  <styled.div
    onClick={[MockFunction]}
  >
    <styled.div
      onClick={[Function]}
    >
      hello, world
      <styled.button
        onClick={[MockFunction]}
        type="button"
      >
        확인
      </styled.button>
    </styled.div>
  </styled.div>
</Portal>
`;
